1. Load and explore the training and test dataset, do necessary pre-processing. 

In [2]:
!pip install imblearn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 258.0/258.0 kB 4.4 MB/s eta 0:00:00a 0:00:01


In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from imblearn.over_sampling import RandomOverSampler

ros = RandomOverSampler(random_state=42)


# Now X_resampled and y_resampled have balanced classes

#load the cirrhosis_train.csv datafile

train=pd.read_csv('data/D_TaskData/cirrhosis_train.csv')
test=pd.read_csv('data/D_TaskData/cirrhosis_test.csv')
        
test_labels=pd.read_csv('data/D_TaskData/sampleprediction.csv')


train.rename(columns={'trainID':'ID'}, inplace=True)
test.rename(columns={'testID':'ID'}, inplace=True)
print(len(train))
print(len(test))

#drop the rows with missing values
train=train.dropna()


224
88


a. Show both training and test dataset size.

In [2]:
#check the dataset size
print(train.shape)
print(test.shape)

(195, 20)
(88, 20)


b. Based on the training and test data, show the feature types, and indicate which 

In [3]:
#print the feature types in the dataset

print(train.dtypes)
print(test.dtypes)

ID                 int64
N_Days             int64
Status            object
Drug              object
Age                int64
Sex               object
Ascites           object
Hepatomegaly      object
Spiders           object
Edema             object
Bilirubin        float64
Cholesterol      float64
Albumin          float64
Copper           float64
Alk_Phos         float64
SGOT             float64
Tryglicerides    float64
Platelets        float64
Prothrombin      float64
Stage              int64
dtype: object
ID                 int64
N_Days             int64
Status           float64
Drug              object
Age                int64
Sex               object
Ascites           object
Hepatomegaly      object
Spiders           object
Edema             object
Bilirubin        float64
Cholesterol      float64
Albumin          float64
Copper             int64
Alk_Phos         float64
SGOT             float64
Tryglicerides    float64
Platelets        float64
Prothrombin      float64
Stage      

c. Use an appropriate method to deal with the missing values for both the training 

d. Do necessary encoding for the categorical features. 

In [4]:
from sklearn.preprocessing import LabelEncoder
#encode categorical values
train=train.apply(LabelEncoder().fit_transform)
test=test.apply(LabelEncoder().fit_transform)


print(train.dtypes)

 

ID               int64
N_Days           int64
Status           int32
Drug             int32
Age              int64
Sex              int32
Ascites          int32
Hepatomegaly     int32
Spiders          int32
Edema            int32
Bilirubin        int64
Cholesterol      int64
Albumin          int64
Copper           int64
Alk_Phos         int64
SGOT             int64
Tryglicerides    int64
Platelets        int64
Prothrombin      int64
Stage            int64
dtype: object


e. Show the label distribution based on the training data, is it a balanced training 
set? 

In [5]:
# Check the column names in the 'train' dataframe
print(train.columns)


# Replace 'class' with the correct column name in the 'train' dataframe
print(train['Status'].value_counts())


Index(['ID', 'N_Days', 'Status', 'Drug', 'Age', 'Sex', 'Ascites',
       'Hepatomegaly', 'Spiders', 'Edema', 'Bilirubin', 'Cholesterol',
       'Albumin', 'Copper', 'Alk_Phos', 'SGOT', 'Tryglicerides', 'Platelets',
       'Prothrombin', 'Stage'],
      dtype='object')
Status
0    102
2     81
1     12
Name: count, dtype: int64


2. Based on the pre-processed training data from question 1, create three supervised 
machine learning (ML) models for predicting “Status”. 

a. Use an appropriate validation method, report performance score using a suitable 
metric. Is it possible that the presented result is an underfitted or overfitted one? 
Justify. 

In [13]:
#create three supervised learning models
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

# Set the parameter grid
# param_grid_svc = {
#     'C': [0.1, 1, 10],  # Regularization parameter
#     'kernel': ['linear', 'rbf'],  # Type of hyperplane
#     'gamma': ['scale', 'auto']  # Kernel coefficient
# }
#split the data into features and target
X_train = train.drop(['ID', 'Status'], axis=1)
y_train = train['Status']

X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

X_train, X_test, y_train, y_test=train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)


# # Create a GridSearchCV object
# grid_search_svc = GridSearchCV(SVC(), param_grid_svc, cv=3, scoring='accuracy', verbose=1)
# grid_search_svc.fit(X_train, y_train)

# # Best parameters and best score
# print("Best parameters:", grid_search_svc.best_params_)
# print("Best cross-validation score: {:.2f}".format(grid_search_svc.best_score_))


# param_grid_lr = {
#     'C': [0.01, 0.1, 1, 10],  # Inverse of regularization strength
#     'solver': ['liblinear', 'saga']  # Algorithm to use in the optimization problem
# }

# grid_search_lr = GridSearchCV(LogisticRegression(max_iter=2000), param_grid_lr, cv=3, scoring='accuracy', verbose=1)
# grid_search_lr.fit(X_train, y_train)

# print("Best parameters:", grid_search_lr.best_params_)
# print("Best cross-validation score: {:.2f}".format(grid_search_lr.best_score_))


# param_grid_rf = {
#     'n_estimators': [100, 200, 300],  # Number of trees in the forest
#     'max_features': ['auto', 'sqrt', 'log2'],  # Number of features to consider at every split
#     'max_depth': [None, 10, 20, 30]  # Maximum number of levels in tree
# }

# grid_search_rf = GridSearchCV(RandomForestClassifier(class_weight='balanced'), param_grid_rf, cv=3, scoring='accuracy', verbose=1)
# grid_search_rf.fit(X_train, y_train)

# print("Best parameters:", grid_search_rf.best_params_)
# print("Best cross-validation score: {:.2f}".format(grid_search_rf.best_score_))


#train the support vector machine model
# Train and evaluate the SVC model
SVC_model = SVC(C=10, gamma='auto', kernel='rbf')
SVC_model.fit(X_train, y_train)
SVC_train_score = SVC_model.score(X_train, y_train)
SVC_test_score = SVC_model.score(X_test, y_test)
SVC_cv_scores = cross_val_score(SVC_model, X_resampled, y_resampled, cv=5)
print("SVC Model:")
print("Train Score:", SVC_train_score)
print("Test Score:", SVC_test_score)
print("Generalization Error (Train-Test Gap):", SVC_train_score - SVC_test_score)
print("Cross-validation Scores:")
print("Mean:", SVC_cv_scores.mean())
print("Standard Deviation:", SVC_cv_scores.std())

# Train and evaluate the Logistic Regression model
LR_model = LogisticRegression(max_iter=2000, C=10, solver='liblinear')
LR_model.fit(X_train, y_train)
LR_train_score = LR_model.score(X_train, y_train)
LR_test_score = LR_model.score(X_test, y_test)
LR_cv_scores = cross_val_score(LR_model, X_resampled, y_resampled, cv=5)
print("\nLogistic Regression Model:")
print("Train Score:", LR_train_score)
print("Test Score:", LR_test_score)
print("Generalization Error (Train-Test Gap):", LR_train_score - LR_test_score)
print("Cross-validation Scores:")
print("Mean:", LR_cv_scores.mean())
print("Standard Deviation:", LR_cv_scores.std())

# Train and evaluate the Random Forest model
RF_model = RandomForestClassifier(class_weight='balanced', max_depth=10, max_features='log2', n_estimators=200)
RF_model.fit(X_train, y_train)
RF_train_score = RF_model.score(X_train, y_train)
RF_test_score = RF_model.score(X_test, y_test)
RF_cv_scores = cross_val_score(RF_model, X_resampled, y_resampled, cv=5)
print("\nRandom Forest Model:")
print("Train Score:", RF_train_score)
print("Test Score:", RF_test_score)
print("Generalization Error (Train-Test Gap):", RF_train_score - RF_test_score)
print("Cross-validation Scores:")
print("Mean:", RF_cv_scores.mean())
print("Standard Deviation:", RF_cv_scores.std())


SVC Model:
Train Score: 1.0
Test Score: 0.8064516129032258
Generalization Error (Train-Test Gap): 0.19354838709677424
Cross-validation Scores:
Mean: 0.7221575885774723
Standard Deviation: 0.034829339267962556

Logistic Regression Model:
Train Score: 0.7213114754098361
Test Score: 0.7419354838709677
Generalization Error (Train-Test Gap): -0.02062400846113166
Cross-validation Scores:
Mean: 0.650290851401375
Standard Deviation: 0.05936271207560896

Random Forest Model:
Train Score: 1.0
Test Score: 0.9354838709677419
Generalization Error (Train-Test Gap): 0.06451612903225812
Cross-validation Scores:
Mean: 0.8889476467477525
Standard Deviation: 0.04999523870882491


SVC Model:

- The model achieves a perfect train score of 1.0, indicating that it fits the training data perfectly.
- However, the test score is lower at 0.8065, suggesting that the model may be overfitting to the training data.
- The generalization error (train-test gap) of 0.1935 is relatively high, further confirming the overfitting issue.
- The mean cross-validation score of 0.7222 is lower than the test score, indicating that the model's performance is not consistent across different folds of the data.
- The standard deviation of cross-validation scores is relatively low at 0.0348, suggesting that the model's performance is somewhat stable across different folds.


Logistic Regression Model:

- The train score of 0.7213 and test score of 0.7419 are relatively close, indicating that the model is not overfitting or underfitting significantly.
- The generalization error (train-test gap) is negative (-0.0206), which means the model performs slightly better on the test data than on the training data. This could be due to random variations or the specific split of the data.
- The mean cross-validation score of 0.6503 is lower than both the train and test scores, suggesting that the model's performance may not be as good when evaluated on different folds of the data.
- The standard deviation of cross-validation scores is relatively high at 0.0594, indicating that the model's performance varies considerably across different folds.


Random Forest Model:

- The model achieves a perfect train score of 1.0, indicating that it fits the training data perfectly.
- The test score of 0.9355 is high, suggesting that the model generalizes well to unseen data.
- The generalization error (train-test gap) of 0.0645 is relatively low, indicating that the model is not severely overfitting.
- The mean cross-validation score of 0.8922 is close to the test score, suggesting that the model's performance is consistent across different folds of the data.
- The standard deviation of cross-validation scores is moderate at 0.0445, indicating some variability in the model's performance across different folds.

b. Justify different design decisions for each ML model used to answer this 
question.

**SVC (Support Vector Classifier):** This model is chosen for its effectiveness in high-dimensional spaces and its capability to model complex nonlinear relationships. The decision to use SVC might be rooted in its ability to find the optimal hyperplane that best separates the data into classes, which is critical in medical or health-related predictions where accuracy is paramount.
**Logistic Regression:** This model is a straightforward, probabilistic approach suitable for binary classification tasks. It’s used here likely due to its simplicity and the interpretability of its results, which is important in clinical settings where understanding the influence of variables is as important as the prediction itself.
**Random Forest:** This ensemble method is utilized for its robustness and ability to handle overfitting, which is common in complex models. By aggregating the results of multiple decision trees, it generally provides high accuracy and handles imbalanced data well, particularly with the use of 'balanced' for class weights to adjust for potential biases in dataset size.

c. Have you optimised any hyper-parameters for each ML model? What are they? 

**SVC (Support Vector Classifier):**


- **C:** Set to 10 to manage the trade-off between achieving a low-error classifier and a decision boundary that is as simple as possible. Higher C values give more flexibility to the model, allowing it to better capture the complexity of the data.
- **gamma:** Kept as 'auto', which automatically adjusts based on the number of features to avoid over-fitting.
- **kernel:** Selected 'rbf' (radial basis function) for its effectiveness in dealing with non-linear boundaries, making the model more adaptive to complex patterns.

**Logistic Regression:**


- **max_iter:** Maintained at 2000 to ensure the algorithm has enough iterations to converge, especially useful when dealing with complex data.
- **C:** Increased to 10, reducing the regularization strength, which allows the model to increase its complexity and potentially improve fit.
- **solver:** Chosen 'liblinear' which is a good choice for small datasets and binary classification, and it handles L1 penalty, useful for feature selection.

**Random Forest:**

- **n_estimators:** Set to 200 to increase the number of trees in the forest, providing a better ensemble model that is less prone to overfitting.
- **max_features:**'log2' is selected, reducing the number of features considered for splitting at each leaf node, which helps in reducing overfitting and increasing the diversity among the trees.
- **max_depth:** Capped at 10 to control the depth of each tree, ensuring the trees do not grow deep enough to fit the noise in the data, which helps in generalizing better.

d. What can you do with the label imbalance issue?

In my implementation I oversampled the lower class and for the random forest I added weighting for the classes

e. Finally, make a model recommendation based on the reported results and justify 
it.

I'd recommend the random forest classifier based on the fact that it has  higher metric scores than the other two models.

3. Use the best model that you get from question 2, do prediction on the pre-processed test 
set. Save your prediction (the prediction should contain two columns only: testID and 
Status), and submit it to the specific Kaggle in-class platform, do a screenshot of your 
model performance and report it. 

In [7]:
# Prepare  test data
X_test = test.drop(['ID', 'Status'], axis=1)

# Predict using the trained model
test_predictions = RF_model.predict(X_test)

# Create a DataFrame with IDs and Predictions
output_df = pd.DataFrame({
    'testID': test['ID']+1,
    'Predicted_Status': test_predictions
})

# Save the DataFrame to a CSV file
output_csv_path = "predicted_status.csv"
output_df.to_csv(output_csv_path, index=False)
print(f"CSV file saved to {output_csv_path}")


CSV file saved to predicted_status.csv
